File: Sources/CCUsageCore/Engine/AnthropicAdapter.swift
import Foundation

/// O que é específico do Claude Code, atrás do `ProviderAdapter`.
///
/// Concentra num lugar só cada coisa que se descobriu observando o Claude Code
/// em agosto/2026 — o nome do item de chaveiro, o `.claude.json` com a
/// identidade, onde ele mora. Nada disto está documentado, então quando uma
/// versão do Claude Code mudar algo, muda-se aqui, e o motor não sente.
public struct AnthropicAdapter: ProviderAdapter {
    public let provider: Provider = .anthropic

    /// Injetável para os testes lerem e escreverem em arquivos temporários.
    private let readFile: @Sendable (URL) -> Data?
    private let writeFile: @Sendable (Data, URL) throws -> Void

    public init(
        readFile: @escaping @Sendable (URL) -> Data? = { try? Data(contentsOf: $0) },
        writeFile: @escaping @Sendable (Data, URL) throws -> Void = { data, url in
            try data.write(to: url, options: .atomic)
        }
    ) {
        self.readFile = readFile
        self.writeFile = writeFile
    }

    public func keychainService(forConfigDir dir: ConfigDir) -> String {
        dir.isDefault
            ? "Claude Code-credentials"
            : "Claude Code-credentials-\(dir.keychainHash)"
    }

    public func identity(inConfigDir dir: ConfigDir) -> AccountIdentity? {
        guard let data = readFile(dir.globalConfigURL),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let oauth = root["oauthAccount"] as? [String: Any],
              let email = oauth["emailAddress"] as? String
        else { return nil }

        return AccountIdentity(
            email: email,
            organizationName: oauth["organizationName"] as? String,
            // O tier fica em `organizationRateLimitTier`; o de usuário é o
            // fallback para conta pessoal sem organização.
            rateLimitTier: (oauth["organizationRateLimitTier"] as? String)
                ?? (oauth["userRateLimitTier"] as? String),
            raw: oauth.mapValues(JSONValue.init(any:)))
    }

    public func writeIdentity(_ identity: AccountIdentity, toConfigDir dir: ConfigDir) throws {
        let url = dir.globalConfigURL
        // Garante o diretório do perfil: um grupo dedicado pode nunca ter sido
        // criado no disco antes da primeira ativação, e escrever o
        // `.claude.json` num diretório inexistente falharia. Para o perfil
        // padrão, o pai é o home, que sempre existe.
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        // Preserva o resto do `.claude.json` — só a identidade e o que é
        // atrelado a ela mudam. Um arquivo ausente vira um objeto novo.
        var root = (readFile(url)
            .flatMap { try? JSONSerialization.jsonObject(with: $0) as? [String: Any] }) ?? [:]

        root["oauthAccount"] = identity.raw.mapValues(\.anyValue)
        // Sem esta chave o Claude Code trata o perfil como primeira execução e
        // abre o assistente de login — sem nem consultar o chaveiro. Identidade
        // escrita = conta já autenticada, então o onboarding está feito.
        root["hasCompletedOnboarding"] = true
        // O cache de uso pertence à conta anterior; deixá-lo carimbaria a nova
        // com números que não são dela até o Claude Code sobrescrever.
        root.removeValue(forKey: "cachedUsageUtilization")

        let data = try JSONSerialization.data(withJSONObject: root,
                                              options: [.prettyPrinted, .sortedKeys])
        try writeFile(data, url)
    }

    public func launchCommand() -> (executable: String, arguments: [String]) {
        ("claude", [])
    }
}

File: windows/crates/router-core/src/engine/anthropic_adapter.rs
//! O que é específico do Claude Code, atrás do [`ProviderAdapter`].
//!
//! Concentra num lugar só o que se descobriu observando o Claude Code — o
//! `.claude.json` com a identidade, onde ele mora, o que precisa estar nele para
//! um perfil dedicado não abrir o assistente de login. Nada disto está
//! documentado; quando uma versão do Claude Code mudar algo, muda-se aqui.

use std::path::PathBuf;

use serde_json::{Map, Value};

use super::account_model::AccountIdentity;
use super::config_dir::ConfigDir;
use super::provider::{IdentityError, Provider, ProviderAdapter};
use crate::platform::atomic_write::read_retrying;
use crate::platform::json_file::{edit_object, JsonFileError};

pub struct AnthropicAdapter;

impl AnthropicAdapter {
    /// A parte pura da leitura, sem I/O — testável com bytes.
    pub fn identity_from_bytes(data: &[u8]) -> Option<AccountIdentity> {
        let root: Value = serde_json::from_slice(data).ok()?;
        let oauth = root.get("oauthAccount")?.as_object()?;
        let email = oauth.get("emailAddress")?.as_str()?.to_string();
        let organization_name = oauth
            .get("organizationName")
            .and_then(Value::as_str)
            .map(String::from);
        // O tier fica em `organizationRateLimitTier`; o de usuário é o fallback
        // para conta pessoal sem organização.
        let rate_limit_tier = oauth
            .get("organizationRateLimitTier")
            .and_then(Value::as_str)
            .or_else(|| oauth.get("userRateLimitTier").and_then(Value::as_str))
            .map(String::from);

        Some(AccountIdentity {
            email,
            organization_name,
            rate_limit_tier,
            raw: oauth.clone(),
        })
    }

    /// O "splice": põe a identidade no objeto raiz do `.claude.json` mexendo só
    /// no que é atrelado à conta. Com o `preserve_order` do `serde_json`, o
    /// `insert` numa chave existente troca o valor **no lugar**, e a remoção é
    /// por `shift` (o `remove` padrão trocaria a última chave de posição).
    pub fn splice_identity(root: &mut Map<String, Value>, identity: &AccountIdentity) {
        root.insert(
            "oauthAccount".to_string(),
            Value::Object(identity.raw.clone()),
        );
        // Sem esta chave o Claude Code trata o perfil como primeira execução e
        // abre o assistente de login — sem nem consultar a credencial. Identidade
        // escrita = conta já autenticada, então o onboarding está feito.
        root.insert("hasCompletedOnboarding".to_string(), Value::Bool(true));
        // O cache de uso pertence à conta anterior; deixá-lo carimbaria a nova
        // com números que não são dela até o Claude Code sobrescrever.
        root.shift_remove("cachedUsageUtilization");
    }
}

impl ProviderAdapter for AnthropicAdapter {
    fn provider(&self) -> Provider {
        Provider::Anthropic
    }

    /// `<perfil>\.credentials.json` — no padrão, `%USERPROFILE%\.claude\.credentials.json`.
    /// Conferido no Windows 11 com o Claude Code 2.1.280 (22/09/2026): arquivo,
    /// nada no Credential Manager. (`CLAUDE_SECURESTORAGE_CONFIG_DIR` desviaria a
    /// leitura para outra pasta; o router a remove de todo processo que lança.)
    fn credential_location(&self, dir: &ConfigDir) -> PathBuf {
        dir.path().join(".credentials.json")
    }

    fn identity(&self, dir: &ConfigDir) -> Option<AccountIdentity> {
        let data = read_retrying(&dir.global_config_path()).ok()?;
        Self::identity_from_bytes(&data)
    }

    /// Grava a identidade de forma **cirúrgica**: lê o `.claude.json` que houver,
    /// troca só `oauthAccount`/`hasCompletedOnboarding`/`cachedUsageUtilization`
    /// e regrava preservando todas as outras chaves, na ordem em que estavam
    /// (inclusive as que só diferem em caixa, que o Windows tem). Um arquivo
    /// existente e ilegível é **recusado**, nunca substituído; ausente ou vazio
    /// vira um objeto novo (grupo dedicado que nunca foi ativado — a pasta do
    /// perfil é criada). Depois de gravar, relê para conferir.
    fn write_identity(
        &self,
        identity: &AccountIdentity,
        dir: &ConfigDir,
    ) -> Result<(), IdentityError> {
        let path = dir.global_config_path();
        edit_object(&path, |root| Self::splice_identity(root, identity)).map_err(|e| match e {
            JsonFileError::Unreadable { path, reason } => {
                IdentityError::Unreadable { path, reason }
            }
            JsonFileError::Io { path, source } => IdentityError::Io { path, source },
        })?;

        match self.identity(dir) {
            Some(back) if back.email == identity.email => Ok(()),
            _ => Err(IdentityError::NotPersisted { path }),
        }
    }

    fn launch_command(&self) -> (String, Vec<String>) {
        ("claude".to_string(), Vec::new())
    }
}
